### PR TITLE
Issue #6 | Add additional settings

### LINKED ISSUE #6: Options | Add additional options for update settings
We should provide more options to user about how the `.json` files should be merged with current settings (currently the only thing users can do is to check if "Groups" names should be updated or not).

Requirements:

- [ ] Move the checkbox "Enable groups" (for `useGroups` option) to the "Search Results Options" column (since this option should exclusively be used for show/hide groups on the context menu, we should have a completely separate option to check if groups names should be updated when refreshing the settings).
- [ ] Rename the label "Last Refreshed on" to "Last Updated on".
- [ ] Rename the "Refresh now" button to "Update now".
- [ ] Rename the "Refresh automatically" label to "Update periodically (once per week)".
- [ ] Add a new title (with is own list of items) called "Merge Settings Options".
- [ ] One merge option should be a dropdown with values "Merge", "Override" and "Ignore" and a side label "Search Providers". Default value for the dropdown should be "merge".
- [ ] Next to the previous label should be a info icon that, when clicked, will open a popover explaining that with "Override" all local search providers will be deleted and replaced when settings are updated, that "Merge" will just add (to the current list of search providers) the search providers that aren't already added from the settings file, and that "Ignore" will keep current list and ignore anything on the settings file.
- [ ] Another merge option should be a checkbox called "Override Groups names" (defaulted to false). 
- [ ] Another merge option should be a checkbox with label "Override Carbon Black Configuration", defaulted to false.
- [ ] Another merge option should be a dropdown with values "Merge", "Override" and "Ignore" and a side label "Carbon Black Queries". Here we should also have a info icon similar to the one of "Search Providers".
- [ ] The finales merge options will be the same checkbox and dropdown from Carbon Black, but for Security Analytics and NetWitness.
- [ ] The values for these merge options should be should also be stored into the Chrome's storage, inside the object from `StoreKey.SETTINGS`, using the field names: `mergeSearchProviders`, `mergeGroups`, `mergeCBC`, `mergeRSA` and `mergeNWI` (these last three with fields `config` and `queries`).
- [ ] Optimize the implementation on `settings.html` use Mustache, by defining a template that will render the checkboxs, inputs and dropdowns (like currently done by the `template_providerConfig` template).
- [ ] On `utils.js` create constants `MERGE_OPTIONS`, `SEARCH_RESULT_OPTIONS` and `CONFIG_FILE_OPTIONS` for define arrays with the data which will be used to generate all inputs from the previous step.
- [ ] Update the `parseJSONFile` function so all these settings are keeping in mind when merging the a JSON file with current settings.
- [ ] Finally, on `migration.js` the value for "Override Groups names" should be initialized to the same value of `useGroups`.

> Please note that, for simplify code when update merge options for CBC / RSA / NWI, you can use the `_.set` and `_.get` methods from Lodash for doing things like `var mergeCBCConfig = _.get( await LocalStore.getOne(StoreKey.SETTINGS), "mergeCBC.config", {})` and `var settings = await LocalStore.getOne(StoreKey.SETTINGS); _.set(settings, "mergeCBC.config", newCBCConfigSettings)`.  
>
> This will also simplify how you define `MERGE_OPTIONS`, since you will be able to do something like:
>
> ```js
> var MERGE_OPTIONS = [
>    { storeKey: "mergeSearchProviders", label: "Search Providers", fields: [...], ... }
>    { storeKey: "mergeGroups", label: "Providers Groups", fields: [...], ... }
>    { storeKey: "mergeCBC.config", label: "Carbon Black Configuration", fields: [...], ... }
>    { storeKey: "mergeCBC.queries", label: "Carbon Black Queries", fields: [...], ... }
>   ...
> ];
> ```

### PR DESCRIPTION
**Pull request by Saidmurod1996**
Fixes #6

### CODE REVIEW

File: src/js/options.js [hunk 1]
```diff
@@ -201,11 +207,35 @@ var SettingsTab = {
...
+          "merge-options"
+        );
+
+        // Enable popovers.
+        $(function () {
+          $('[data-toggle="popover"]').popover({
+            trigger: "hover",
+            html: true,
+          });
+        });
```
Comment (jfmdev) [inline src/js/options.js:237]: Does the popover's implementation from [Boostrap.native](http://thednp.github.io/bootstrap.native/#componentPopover) isn't good?

> We originally removed jQuery because: it wasn't really need; it opens the door to bad programming practices; for security (isn't that jQuery is itself insecure, but it's targeted by hackers).  
> When implementing the modal for edit the JSON settings, Boostrap.Native was added to _avoid_ having to re-introduce jQuery.  
> However if Bootstrap.Native isn't doing his job (for example: don't works correctly for popovers), then we should remove it and use only jQuery, but we shouldn't use **both** jQuery and Bootstrap.Native.

File: src/views/settings.html [hunk 2]
```diff
@@ -3,168 +3,155 @@
...
+                <div>
+                  <label class="mb-1">{{label}}</label>
+                  <div id="settings_{{key}}"></div>
+                  <div
+                    class="text-danger"
+                    id="settings_{{key}}Error"
+                  ></div>
+                </div>
+
+              {{/isText}}
```
Comment (jfmdev) [inline src/views/settings.html:63]: If `lastConfigUpdate` is the **only** item of type `text`, then maybe we shouldn't include it on the Items loops. We can just put it outside, like (note this is pseudo-code):

```
{{/items}}

<div>
  <label class="mb-1">Last Refreshed on:</label>
   <div>{{ lastConfigUpdate || "Never" }}</div>
     {{#lastConfigUpdateError}}
         <div class="text-error">{{ lastConfigUpdateError }}</div>
     {{/lastConfigUpdateError}}
</div>
<div>
  <button
    type="button"
    class="btn btn-success"
    id="settings_refreshNow"
  >
    <i class="fas fa-sync"></i> Refresh Now
  </button>
</div>
```

File: src/views/settings.html [hunk 3]
```diff
@@ -3,168 +3,155 @@
...
+                            class="form-check-input"
+                            name="{{storeKey}}"
+                            id="settings_{{storeKey}}"
+                    />
+                    <label class="form-check-label" for="settings_{{storeKey}}">
+                      {{label}}
+                    </label>
+                  </div>
+                </li>
+              {{/isCheckbox}}
```
Comment (jfmdev) [inline src/views/settings.html:129]: Note that you can define a Template at the bottom of the page (or at any part of the page) a then use it to _populate_ several parts of the page. Like:

```
var genericTemplate = document.getElementById("some-template")).innerHTML;

var sectionOne = Mustache.render(genericTemplate , dataForSection1);
document.getElementById("section_one").innerHTML = sectionOne;

var sectionTwo = Mustache.render(genericTemplate , dataForSection2);
document.getElementById("section_two").innerHTML = sectionTwo;
```

> I'm mentioning this this because `template_merge-options` and `template_config` have copy/pasted code between them.
>
> Also note that this is what we currently do for CBC, RCA and NWI: we defined two generic templates (one for configuration values and other for queries) that we later uses on each of these three tabs.

**Review by jfmdev — CHANGES_REQUESTED** [review]

File: src/js/options.js (hunk 1 above)
Comment (Saidmurod1996) [reply src/js/options.js]: I've used Bootstrap.native for dropdown and popover and removed jquery

**Comment by pushpendraKh** [comment]
you added `.idea` files too in your last commit. @Saidmurod1996

**Comment by pushpendraKh** [comment]
Functionality-wise working as expected. 

- [ ] Dropdown does not dismiss once an item is selected.

Some changes I would suggest - 
- [ ] Show the popover in the right, if possible
- [ ] Content in dropdown should be in order -> Merge, Override and Ignore
- [ ] Items in the `Merge Settings options` can be in the order of TABs. This means, Security Analytics, Netwitness, followed by CBC (override config and query. )

**Comment by pushpendraKh** [comment]
@Saidmurod1996 still forgot to remove `.idea` folder (add it to gitignore also)

**Comment by pushpendraKh** [comment]
@Saidmurod1996 

- [ ] Dropdown still does not dismiss once an item is selected. Check CBC dropdown. Here is console error.

dropdown_error

- [ ] You have used tooltip but its background color is so black that it is not looking good. Maybe less opacity or garyish may look good.  Plus some margin on left when you are showing the tooltip.

tooltip

File: src/css/main.css [hunk 4]
```diff
@@ -56,3 +56,7 @@ html.page-options h3 {
 .sortable.sortable-ghost {
   background-color: #DDF;
 }
+
+.dropdown button {
+  border-color: grey;
+}
```
Comment (jfmdev) [inline src/css/main.css:62]: No need for this CSS change, for now we'll stick to colors defined by Bootstrap

**Comment by jfmdev** [comment]
No need to change the color of the tooltip neither (although making the margin more consistent would be good), we'll update/improve colors on #4

**Comment by pushpendraKh** [comment]
@jfmdev @Saidmurod1996  
- Tooltip is no longer working on hover or on click. 
- After selecting an option from the dropdown, it lefts a shadow.

Screenshot 2021-06-05 at 20 13 01

**Comment by jfmdev** [comment]
The shadow is due to the element being focused, hence it's expected, and the tooltip should be fixed after I commit a change

File: src/js/options.js [hunk 5]
```diff
@@ -124,7 +128,7 @@ var Header = {
     },
   ],
 
-  DEFAULT_TAB: "search-providers",
+  DEFAULT_TAB: "settings",
```
Comment (pushpendraKh) [inline src/js/options.js:131]: Undo this change.

File: src/js/options.js [hunk 6]
```diff
@@ -258,7 +321,23 @@ var SettingsTab = {
       await LocalStore.setOne(StoreKey.SETTINGS, newSettings);
 
       // Update context menu
-      mainConfigurationUpdated(true);
+      mainConfigurationUpdated();
+    }
+  },
+
+  onDropdownSelect: async function (event) {
+    console.log(event);
```
Comment (pushpendraKh) [inline src/js/options.js:329]: Remove log.

**Comment by pushpendraKh** [comment]
@Saidmurod1996 @jfmdev 

I didn't notice it earlier.. 

- We are showing the same tooltip message for all the dropdown buttons, which holds true for search providers. 
- As stated before, Good to have a margin before tooltip.

The functionality works fine as before.

File: src/js/options.js [hunk 7]
```diff
@@ -258,7 +321,23 @@ var SettingsTab = {
       await LocalStore.setOne(StoreKey.SETTINGS, newSettings);
 
       // Update context menu
-      mainConfigurationUpdated(true);
+      mainConfigurationUpdated();
```
Comment (jfmdev) [inline src/js/options.js:324]: Why we must refresh **all** settings after an input change?

File: src/js/options.js [hunk 8]
```diff
@@ -258,7 +321,23 @@ var SettingsTab = {
...
+    var targetName = _.get(event, "target.name");
+
+    if (!_.isEmpty(targetName)) {
+      var newSettings =
+        _.clone(await LocalStore.getOne(StoreKey.SETTINGS)) || {};
+      newSettings[targetName] = event.target.value;
+      await LocalStore.setOne(StoreKey.SETTINGS, newSettings);
+
+      // Update The UI
+      SettingsTab.updateForms();
```
Comment (jfmdev) [inline src/js/options.js:340]: Same as above, why we must re-render all forms on the Settings tabs after a dropdown is changed?

File: src/js/options.js [hunk 9]
```diff
@@ -201,11 +205,37 @@ var SettingsTab = {
...
         // Insert template file.
         document.querySelector('main section[data-tab="settings"]').innerHTML =
           htmlData;
 
+        await SettingsTab.injectData(CONFIG_FILE_OPTIONS, "config");
+        await SettingsTab.injectData(SEARCH_RESULT_OPTIONS, "search-results");
+        await SettingsTab.injectData(MERGE_OPTIONS, "merge-options");
+
+        // Initialize tooltips.
+        let popovers = document.querySelectorAll("[data-title]");
```
Comment (jfmdev) [inline src/js/options.js:218]: This selector is to _broad_: it's targeting all elements on the page that have `data-title` (right now we only have tooltips here, but this will cause troubles if we add tooltips or popovers elsewhere), the selector should be only affecting elements on the settings tab. something like `document.querySelectorAll('main section[data-tab="settings"] [data-title]')`

File: src/js/shared/config_file.js [hunk 10]
```diff
@@ -153,26 +153,46 @@ const ConfigFile = {
...
-      if (
-        !_.find(searchProviders, function (provider) {
-          return provider.link === newProviders[i].link;
-        })
-      ) {
-        searchProviders.push(newProviders[i]);
+
+    // - Check if the new list of search providers are already included on the list of menu items.
+    var mergeProviderOption = _.get(settings, "mergeSearchProviders", "0");
+    if (mergeProviderOption === "0") {
```
Comment (jfmdev) [inline src/js/shared/config_file.js:160]: Using numbers as values isn't very _dev-friendly_, `mergeProviderOption ` should have values "merge", "override" and "ignore" instead of "0", "1", "2"

File: src/js/shared/config_file.js [hunk 11]
```diff
@@ -153,26 +153,46 @@ const ConfigFile = {
...
+          })
+        ) {
+          searchProviders.push(newProviders[i]);
+        }
       }
+    } else if (mergeProviderOption === "1") {
+      // override
+      searchProviders = newProviders;
+    } else {
+      // ignore
```
Comment (jfmdev) [inline src/js/shared/config_file.js:176]: This `else` can be removed, since nothing is done on this block

File: src/js/shared/config_file.js [hunk 12]
```diff
@@ -212,11 +232,44 @@ const ConfigFile = {
...
+    var shouldOverrideConfig = _.get(settings, mergeKey + ".config", false);
+    var queriesMergeOption = _.get(settings, mergeKey + ".queries", "0");
+
+    var provData = (await LocalStore.getOne(storeKey)) || {};
+
+    if (shouldOverrideConfig) {
+      provData.config = _.get(newData, "Config", {});
+    }
+
+    if (queriesMergeOption === "0") {
```
Comment (jfmdev) [inline src/js/shared/config_file.js:246]: Same as above, `queriesMergeOption ` should contains strings, instead of numbers (so it's more clear to other devs what each value means).

File: src/js/shared/constants.js [hunk 13]
```diff
@@ -112,6 +128,55 @@ export const NWI_CONFIG = [
   },
 ];
 
+export const MERGE_DROPDOWN_ITEMS = ["Merge", "Override", "Ignore"];
```
Comment (jfmdev) [inline src/js/shared/constants.js:131]: Due to my comments above, maybe this should be changed to: `MERGE_DROPDOWN_ITEMS = [{key: 'merge', label: "Merge"}, {key: 'override', value: "Override"}, ...];`

**Review by jfmdev — CHANGES_REQUESTED** [review]
Just a few comments (we're almost there  🙂 )

File: src/js/shared/config_file.js [hunk 14]
```diff
@@ -212,11 +230,44 @@ const ConfigFile = {
...
+        }
+      }
+    } else if (queriesMergeOption === "override") {
+      // override
+      provData.queries = _.map(
+        _.get(newData, "Queries", []),
+        ConfigFile.parseQuery
+      );
+    } else {
+      // ignore
```
Comment (jfmdev) [inline src/js/shared/config_file.js:267]: This `else` block is unnecessary

**Review by jfmdev — CHANGES_REQUESTED** [review]
The last requirement wasn't implemented: `on migration.js the value for "Override Groups names" should be initialized to the same value of useGroups.`

**Review by jfmdev — APPROVED** [review]
